tools/update_snapcraft.py

```diff
@@ -75,9 +79,58 @@ def get_latest_tempestconf_requirements():
     return OPENINFRA_REPO_URL_FMT.format(project=project, ref=latest_revision)
 
 
+def clone_releases_repository():
+    """Clone the OpenStack releases repository.
+
+    If the repository exists, just fetch origin and fast-forward
+    to the tip of the current remote default branch.
+
+    If fast forwarding not possible, e.g. repo is dirty.
+    Remove the repo and reclone.
+    """
+    RELEASES_REPO_PATH.parent.mkdir(parents=True, exist_ok=True)
+    try:
+        pygit2.clone_repository(RELEASES_REPO_URL, RELEASES_REPO_PATH)
+    except ValueError as error:
+        repo_path = pygit2.discover_repository(RELEASES_REPO_PATH)
+        if not repo_path:
+            raise RuntimeError(
+                f"{RELEASES_REPO_PATH} is not empty and not a git repository"
+            ) from error
+
+        repo = pygit2.Repository(repo_path)
+        repo.checkout("refs/heads/master")
+        repo.remotes["origin"].fetch()
+        remote_master_id = repo.lookup_reference("refs/remotes/origin/master").target
+        analysis_result, _ = repo.merge_analysis(remote_master_id)
+
+        if analysis_result & pygit2.GIT_MERGE_ANALYSIS_UP_TO_DATE:
+            logger.info("Release repository up to date")
+            return
+
+        if analysis_result & pygit2.GIT_MERGE_ANALYSIS_FASTFORWARD:
+            # pygit2 does not expose a simple way to perform a ff merge
+            # the following is a manual ff merge:
+            # checkout to the remote master reference object
+            # then update the local master to match it
+            logger.info("Updating release repository")
+            repo.checkout_tree(repo.get(remote_master_id))
+            try:
+                master_ref = repo.lookup_reference("refs/heads/master")
+                master_ref.set_target(remote_master_id)
+            except KeyError:
+                repo.create_branch("master", repo.get(remote_master_id))
+            repo.head.set_target(remote_master_id)
+        else:
+            # ff merge not available. We might as well reclone. it will be messy otherwise
+            logger.info("Release repository not clean. Recloning.")
+            shutil.rmtree(RELEASES_REPO_PATH)
+            pygit2.clone_repository(RELEASES_REPO_URL, RELEASES_REPO_PATH)
```

Comment:
Personally I think this function is a bit overkill. Why can't we simply remove the cloned directory upon completion of this python script? Then we can have sequential runs, and if the path already exist when cloning the directory (meaning that there's something unknown occupying the directory and we probably shouldn't blindly overwrite it) we can just exit in error.

Adding unnecessary code increases future maintenance burden, so I prefer to keep the logic as simple as possible.

Reply:
That was my first instinct as well since i also did not want to write logic for the repo updates. However upon checking the releases repository, I realized its size is 47 MB as of today. When combined with the fact that [pygit2 does not have a `depth` parameter in `clone_repository()`](https://github.com/libgit2/pygit2/issues/630),  

we are left with:
- this approach, which I don't personally find unnecessary or overkill given the circumstances
- subprocess calls to the git cli to clone with depth 1(still comes to 25 MB)
- giving up on local usage feature of the script altogether
- pulling 47 MB with each invocation
- other solutions that don't come to my mind

I am personally at equal distance to any of the above. So I leave it to the maintainers to decide at this point. Will happily update the PR with suggestions.

Reply:
I see, thanks for the explanation. 

Aside from having complex logic, the part of this solution I worry about the most is blindly overwriting the path if it already exists and it's a git directory (we don't know the actual contents and it could be tracking other remotes). Though I know the chances are low and it only happens during local usage, I feel like we should avoid this risk. If we end up going with this approach, we should at least check the remote origin of the repo and see if it matches RELEASES_REPO_URL before trying to do at operations with it.

A second opinion is appreciated.

Reply:
yeah, I agree. that check must be here. if the decision is to keep this solution, I will definitely add that check. thanks for that!  
maybe @esunar or @aieri could give a second opinion about how to move forward about this particular bit here.

Reply:
tbh I don't expect local usage to be a frequent use case. For this reason, I'd even be ok with the very simple old behavior.
The releases repo being 47 MB is not great though, and not having `--depth` in pygit2 makes it even worse. On the other hand, this script will most of the time be run within the CI.

My proposal would be:
* use `tempfile` instead of `/tmp/releases`
* clean up when the script is done, unless a CLI option is passed to prevent it
* subscribe to https://github.com/libgit2/pygit2/issues/630 and update this code when a depth parameter is added

Reply:
Since we are defining the tox testenv, the git clone command can be used before calling this script. It's just an idea, not saying we must do it. Since most time this will be run only with CI it's ok.
One more think, it will be faster since CI will not install any more Python dependencies.

Reply:
for simplicity, in the update I added a flag called `--reuse` and added `tempfile` usage, which I have missed in the first iteration.